### PR TITLE
Build command implementation for SVN material

### PR DESCRIPTION
**Pull request by wpc**
This PR is part of the #1954 build command protocol proposal, and a piece of work following PR #2052.
### Scope:
- [x] Make condition logic of build command language nicer, make compose robust logic on server side easier
- [x] Support svn material, porting related test

### CODE REVIEW

**Comment by wpc** [comment]
@arvindsv  This PR is ready for review.

**Comment by arvindsv** [comment]
I can see you're enjoying yourself. You're implementing lisp. :)

File: common/test/unit/com/thoughtworks/go/buildsession/BuildSessionCancelingTest.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public void cancelLongRunningTestCommand() throws InterruptedException {
             @Override
             public void run() {
                 buildSession.build(compose(
-                        echo("after sleep").setTest(execSleepScript(50))));
+                        cond(or(execSleepScript(50), execSleepScript(50)), echo("after sleep"))));
```
Comment (arvindsv) [inline common/test/unit/com/thoughtworks/go/buildsession/BuildSessionCancelingTest.java]: Why two of these now?

Comment (wpc) [inline common/test/unit/com/thoughtworks/go/buildsession/BuildSessionCancelingTest.java]: to make sure testing command after testing command being canceled will not execute.

File: common/src/com/thoughtworks/go/buildsession/ExecCommandExecutor.java [hunk 2]
```diff
@@ -85,7 +91,7 @@ private CommandLine createCommandLine(String cmd) {
         return commandLine;
     }
 
-    private int executeCommandLine(final BuildSession buildSession, final CommandLine commandLine) {
+    private int executeCommandLine(final BuildSession buildSession, final CommandLine commandLine, boolean verbose) {
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/buildsession/ExecCommandExecutor.java]: Was executeCommandLine supposed to use this new boolean? It is not.

File: common/test/unit/com/thoughtworks/go/buildsession/ExecCommandExecutorTest.java [hunk 3]
```diff
@@ -133,6 +134,12 @@ public void shouldNotLeakSecretsToLog() {
         }
     }
 
+    @Test
+    public void shouldReportErrorWhenFailExecWithVerbose() {
+        runBuild(exec("not-not-not-exist", true, "-a", "-bb"), Failed);
+        System.out.println("console = " + console);
```
Comment (arvindsv) [inline common/test/unit/com/thoughtworks/go/buildsession/ExecCommandExecutorTest.java]: Is there an assert missing here?

File: common/test/unit/com/thoughtworks/go/buildsession/BuildSessionCancelingTest.java (hunk 1 above)
Comment (arvindsv) [inline common/test/unit/com/thoughtworks/go/buildsession/BuildSessionCancelingTest.java]: Ok.

File: common/src/com/thoughtworks/go/domain/materials/svn/SvnMaterialUpdater.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    private BuildCommand cleanupAndUpdate(String workingDir, Revision revision) {
+        return compose(
+                svn("cleanup", workingDir),
+                svn("revert", "--recursive", workingDir),
+                svnWithAuth("update", "--non-interactive", "-r", revision.getRevision(), workingDir));
+    }
+
+    private BuildCommand isRepositoryChanged(String workingDir) {
+        return test("-ncontains",
+                "URL: " + StringUtil.removeTrailingSlash(material.getUrl()),
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/domain/materials/svn/SvnMaterialUpdater.java]: I think there's a newline missing here. It fails this test:

```
@Test
public void shouldCheckoutIfSvnRepositoryChangesToAParentRepo() throws IOException {
    String initialRepo = testRepo.projectRepositoryUrl();
    runBuild(buildCommand("3", initialRepo), JobResult.Passed);
    assertThat(console.output(), containsString("Checked out revision 3"));

    String changedRepo = FileUtil.toFileURI(testRepo.projectRepositoryRoot());
    runBuild(buildCommand("4", changedRepo), JobResult.Passed);
    assertThat(console.output(), containsString("Checked out revision 4"));

    assertThat(getWorkingDirInfo(workingCopy).getUrl() + "/", is(changedRepo));
}
```

If that is right, you should probably take a look at removing that `+ "/"` in the test.

**Comment by arvindsv** [comment]
Everything else looks good to me.

File: common/test/unit/com/thoughtworks/go/buildsession/ExecCommandExecutorTest.java (hunk 3 above)
Comment (wpc) [inline common/test/unit/com/thoughtworks/go/buildsession/ExecCommandExecutorTest.java]: Oops. Sorry about that.

File: common/src/com/thoughtworks/go/buildsession/ExecCommandExecutor.java (hunk 2 above)
Comment (wpc) [inline common/src/com/thoughtworks/go/buildsession/ExecCommandExecutor.java]: Not really. Removing it

**Comment by wpc** [comment]
@arvindsv Finally sometime to finish up this PR. sorry for the delay.

**Comment by wpc** [comment]
@arvindsv I added a "script" flag to exec command, for making behavior same as old code on windows. Also it is helpful for getting consistent result cross platform in the tests.

**Comment by xli** [comment]
@ketan @arvindsv this is read for merge, there is failure on build but it's build environment related.

**Comment by ketan** [comment]
Closing PR since there's a lot of merge conflicts, and not enough interest.

**Comment by ketan** [comment]
Closing PR since there's a lot of merge conflicts, and not enough interest.

**Pull request closed without merging**